.github/workflows/ci.yml

```
@@ -24,7 +24,13 @@ jobs:
       - name: Setup Python
         uses: actions/setup-python@v4
         with:
-          python-version: 3.9
+          python-version: 3.12
+      - name: Update pip
+        run: |
+          python -m pip install -U pip
+      - name: Install requirements
+        run: |
+          python -m pip install -U -r requirements.txt -r tests/requirements.txt -r requirements-mypy.txt
```

I think we can zip both steps together

```suggestion
      - name: Install requirements
        run: |
          python -m pip install -U pip
          python -m pip install -U -r requirements.txt -r tests/requirements.txt -r requirements-mypy.txt
```

---

If you choose to do it, you can do the same for `test` job.